aten/src/ATen/Device.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include <ATen/Error.h>
+#include <ATen/ScalarType.h>
+#include <ATen/Tensor.h>
+#include <ATen/TensorMethods.h>
+#include <ATen/Type.h>
+#include <ATen/optional.h>
+
+#include <cstddef>
+#include <iosfwd>
+#include <string>
+
+namespace at {
+/// Represents a a compute device on which a tensor is located. A device is
+/// uniquely identified by a type, which specifies the type of machine it is
+/// (e.g. CPU or CUDA GPU), and a device index or ordinal, which identifies the
+/// specific compute device when there is more than one of a certain type. The
+/// device index is optional, and in its defaulted state represents (abstractly)
+/// "the current device". Further, there are two constraints on the value of the
+/// device index, if one is explicitly stored:
+/// 1. The device index is never negative.
+/// 2. When the device type is CPU, the device index must be zero.
+struct Device {
+  /// The possible values of the device *type*.
+  enum class Type { CPU, CUDA };
+
+  /// Converts a `Backend` to a `Device::Type` if possible.
+  static Type backend_to_type(Backend backend) {
+    switch (backend) {
+      case kCPU:
+      case kSparseCPU:
+        return Type::CPU;
+      case kCUDA:
+      case kSparseCUDA:
+        return Type::CUDA;
+      default:
+        AT_ERROR(
+            "Invalid backend ", toString(backend), " for Device construction");
+    }
+  }
+
+  /// Constructs a new `Device` from a `Type` and an optional device index.
+  /* implicit */ Device(Type type, at::optional<int32_t> index);
+  /* implicit */ Device(Type type);
+
+  /// Constructs a `Device` from a string description, for convenience.
+  /// The string supplied must follow the following schema:
+  /// `[(cpu|cuda):][<device-index>]`
+  /// where `cpu:` or `cuda:` optionally specify the device type, and
+  /// `<device-index>` optionally specifies a device index. The device index
+  /// defaults to `nullopt`. If no device index is given, the device
+  /// type defaults to CPU. If a device index is supplied, the type defaults to
+  /// CUDA, such that a string like "1" would be equivalent to "cuda:1".
+  /* implicit */ Device(const std::string& device_string);
```

Comment:
do we need this in the API?  I put something similar in the pytorch API for backwards compatibility reasons, but we don't really have devices in ATen yet, so that doesn't seem to be a concern.  Can we just bridge where we need it in pytorch?

Reply:
I thought it would be nice for research users. It's not connected to the Python implementation

Reply:
Oh I see you mean the behavior with only a digit

Reply:
Good point, I should remove that

Reply:
right, sorry for not being clear.  I expected just an integer to be self contained at the PyTorch level.